Marlin/Marlin_main.cpp

```diff
@@ -4053,6 +4054,13 @@ inline void gcode_M109() {
 #endif // HAS_TEMP_BED
 
 /**
+ * M110: Set Current Line Number
+ */
+inline void gcode_M110() {
+  if (code_seen('N')) gcode_N = code_value_long();
```

Comment:
This is a pretty clever way to work it out, although `gcode_N` will end up being set twice when there's an `N` parameter in front. This can actually be avoided by adding a little code here:
```cpp
if (M110) {
  char* n2pos = strchr(command + 4, 'N');
  if (n2pos) npos = n2pos;
}
```
… to this …
```cpp
if (M110) {
  char* n2pos = strchr(command + 4, 'N');
  if (n2pos) {
    npos = n2pos;
    *npos = '\0'; // don't process this 'N' again!
  }
}
```
There's also this line that starts `else if (npos == command) {` and it's bugging me. It seems that it's meant to make sure that the last 'N' parameter found is the first one, and only complain about no checksum in that case. But I don't think it's needed any more. It used to be needed because `npos` could be found in the wrong place. But now we only set `npos` initially from the leading 'N'. Long story short, that line line should just be `else {` now.

Anyway, this parser leaves much to be desired, so maybe it will get replaced a parser from one of the other firmwares that I have been studying. I like the ones that scan all the parameters up front and store them before ever calling the command handlers. Some are very smart, even making sure not to do any floating point maths. Even though the command parser is not our bottle-neck, it certainly isn't the most efficient, and it's from old Sprinter code, so I know we can do better.